docker trust: interact with signers and keys

riyazdf opened the pull request:
This PR introduces four more `docker trust` subcommands to interact with signers and keys:

- `docker trust signer add`: add a signer to one or more repos
- `docker trust signer remove`: remove a signer from one or more repos
- `docker trust key load`: load a private key into the local storage to use for signing
- `docker trust key generate`: generate a signing key-pair.  The private key is automatically loaded into the local storage, and the public key can be used for `signer add`

<img src="https://images.fineartamerica.com/images-medium-large/cat-and-lock-dennis-glowacki.jpg" width="400"></img>

cc @ashfall and @eiais who also implemented these features

This PR includes code and documentation and extends the functionality in #472. As such, it can be included in a followup release.

[Original design doc here](https://docs.google.com/document/d/1JOBAlCDuf5JnpVLW54voGuAdsnMmSR2LIbs8fjBDSaM/edit )

thaJeztah commented:
Just from the description; should we think of making `signer` and `key` top-level objects, to make it fit better with the existing UX (`docker signer add`, `docker signer remove`, `docker key load`, `docker key generate`), or alternatively `docker trust-key load`, `docker trust-key generate`

riyazdf commented:
@thaJeztah: we kept `key` and `signer` under the `trust` command because the concepts probably won't be used elsewhere - that said I'm cool with factoring them out to `docker signer`/`docker key` or `docker trust-signer`/`docker trust-key` if you think that would be better.

For most users the flows will look something like:

```
# Add a signer and view to confirm
docker trust signer-add alice repo
docker trust view repo
```

```
# Load my signer key and sign with it
docker trust key-load alice.priv
docker trust sign repo
```

dnephin commented:
I like it either as it is now, or as another level under `docker trust` `docker trust signer add` , `docker trust key load`. 

Moving them to the top level would be confusing.

thaJeztah commented:
```
docker trust signer add
```

works for me; would that be more difficult to implement (as we currently don't have sub-sub commands)?

dnephin commented:
Shouldn't be. All we need is a couple additional `cobra.Command` objects for `signer` and `key` and add the new commands to them instead of the `Command` for trust.

riyazdf commented:
@dnephin @thaJeztah : sounds good!  Are we okay with these being one of the only (if not the only) commands in `docker/cli` with four levels of subcommands?

If there's consensus I'm happy to make the change.

thaJeztah commented:
I'm OK with it; it may be slightly awkward to have the commands nested this deep, but with the context given above, I think it's the best option (and most consistent with other commands as in `<something>  add`)

ping @docker/core-cli-maintainers PTAL

albers commented:
The completions are not equipped for another nesting level, so I'd prefer the less deeply nested forms.

vieux commented:
I'm ok with `docker trust signer add` and `docker trust key load`

We almost went for `docker swarm node update` instead of a top level`docker node` so I'm not shocked by this depth

riyazdf commented:
Ok!  From discussions with @dnephin, @thaJeztah, @vdemeester, @stevvooe, and @vieux : it sounds like `docker trust signer add`et al. is the best path forward.  I will update this PR shortly.

I'm not sure what the implications are for bash completions, is it possible to make it work with this level deep of commands?

albers commented:
For bash completion it will be 0.5 to 3 days of work.
The current implementation uses global variables for the command and subcommand positions. It might be hard to turn this into a proper generic solution.

I'm also not sure about the implications on zsh and fish completion.
@sdurrheimer can zsh completion deal with an additional level of subcommands?

riyazdf commented:
@albers: got it.

I've moved the commands around to be a level lower, now we have:
```
🐳 $ ./docker-darwin-amd64 trust

Usage:	docker trust COMMAND

Manage trust on Docker images (experimental)

Options:
      --help   Print usage

Management Commands:
  key         Manage keys for signing Docker images (experimental)
  signer      Manage entities who can sign Docker images (experimental)

Commands:
  revoke      Remove trust for an image
  sign        Sign an image
  view        Display detailed information about keys and signatures

Run 'docker trust COMMAND --help' for more information on a command.
```

dnephin commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,139 @@
+	if err != nil {
+		return err
+	}
+	addStagedSigner(notaryRepo, newSignerRoleName, signerPubKeys)
```
Not related to this PR, but I just took a look at `addStagedSigner()` and I noticed that it calls 4 interface methods that all return an error, but none of the errors are checked. Is that safe? Does the implementation not return important errors?

dnephin commented on cli/command/trust/signer_remove.go:
```diff
@@ -0,0 +1,137 @@
+		Short: "Remove a signer",
+		Args:  cli.RequiresMinArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return removeSigner(dockerCli, args[0], args[1:], &options)
```
Most commands (including `newSignerAddCommand` above) pass args using the options struct instead of separate args. I think it would be good to do that here as well.

Also it shouldn't be necessary to use a pointer for options. The values should never be modified, so I believe we pass a struct (not a pointer) in all other commands.

dnephin commented on cli/command/trust/signer_remove.go:
```diff
@@ -0,0 +1,137 @@
+		},
+	}
+	flags := cmd.Flags()
+	flags.BoolVarP(&options.forceYes, "yes", "y", false, "Answer yes to removing most recent signer (no confirmation)")
```
This should probably say something like "do not prompt for confirmation before ..." or something along those lines

dnephin commented on cli/command/trust/signer_remove.go:
```diff
@@ -0,0 +1,137 @@
+	return counter < releasesRoleWithSigs.Threshold, nil
+}
+
+func removeSingleSigner(cli command.Cli, image, signerName string, forceYes bool) error {
```
When `signerName` is part of options I think `signerName` and `forceYes` can be replaced with `options signerRemoveOptions`

riyazdf replied:
I left this function signature as is because otherwise there's a weird discrepancy between the unpacked image named and otherwise packed options.

dnephin replied:
I don't really see how this is a problem. It's quite common for a function to only use some fields or methods on a struct.

riyazdf replied:
@dnephin: I don't think that is a safe practice: this is an inner helper function that only needs to know about one repo.  Including the extra images in the slice is extraneous info, and I worry that a future contribution may misuse it.

That said, if you feel strongly about changing to the `signerRemoveOptions`, you are the maintainer and I will make the change - just let me know.

dnephin commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,139 @@
+
+var validSignerName = regexp.MustCompile(`^[a-z0-9]+[a-z0-9\_\-]*$`).MatchString
+
+func addSigner(cli command.Cli, options *signerAddOptions) error {
```
should accept the struct, not pointer to struct

dnephin commented on vendor.conf:
```diff
@@ -1,5 +1,9 @@
+github.com/bugsnag/bugsnag-go 13fd6b8acda029830ef9904df6b63be0a83369d0
+github.com/bugsnag/osext 0dd3f918b21bec95ace9dc86c7e70266cfc5c702
+github.com/bugsnag/panicwrap e2c28503fcd0675329da73bf48b33404db873782
+github.com/BurntSushi/toml a368813c5e648fee92e5f6c30e3944ff9d5e8895
```
This is bringing in a whole lot of unnecessary dependencies.

I believe the problem is `github.com/docker/notary/util`. I think this illustrates one of the major problems with a `utils` package. 

from https://blog.golang.org/package-names

> Avoid meaningless package names. Packages named util ...

A generic package like `utils` tends to collect a bunch of unrelated functions/types that have completely different dependencies which means that using any of the functions results in the consumer having to depend on a bunch of unrelated things (sql libraryes, logging libraries, etc).

I think this needs to be fixed in notary and anything used in `docker/cli` needs to be moved out of `utils` so we can use it without bringing in a dozen unnecessary packages.

riyazdf replied:
I'm looking into it - it's possible that this may not be so clean because the notary server-side services also use a key storage interface with connects with the DB (hence many of the imports).  I'll get back to you with an update.

riyazdf commented on cli/command/trust/key_generate.go:
```diff
@@ -0,0 +1,107 @@
+func setupPassphraseAndGenerateKeys(streams command.Streams, keyName string) error {
+	// always use a fresh passphrase for each key generation
+	freshPassRetGetter := func() notary.PassRetriever { return trust.GetPassphraseRetriever(streams.In(), streams.Out()) }
+	cwd, err := os.Getwd()
```
agreed!  Added.

riyazdf commented:
thanks @dnephin for the review!  I'm looking into package refactors on `docker/notary` and have applied your other suggestions :)

riyazdf commented:
@dnephin: great news!  I have a fix in flight for upstream notary that will remove all vendoring except for an updated notary commit hash.

We would love to move this forward before the next release - please let me know if there are other review comments I can address :)

riyazdf commented:
updated: the only vendoring is now limited to an updated `docker/notary` commit sha.  Excited to move forward with this!

dnephin commented on cli/command/trust/key_generate.go:
```diff
@@ -0,0 +1,130 @@
+	pubPEM, err := generateKeyAndOutputPubPEM(keyName, privKeyFileStore)
+	if err != nil {
+		fmt.Fprintf(streams.Out(), err.Error())
+		return fmt.Errorf("Error generating key for: %s", keyName)
```
I don't think it's correct to print an error on stdout. This should return an error without the `fmt.Fprintf()`

```go
return errors.Wrapf(err, "failed to generate key for %s", keyName)
```

dnephin commented on cli/command/trust/key_generate.go:
```diff
@@ -0,0 +1,130 @@
+	pubFileName := strings.Join([]string{keyName, "pub"}, ".")
+	pubFilePath := filepath.Join(workingDir, pubFileName)
+	if err := ioutil.WriteFile(pubFilePath, pem.EncodeToMemory(&pubPEM), notary.PrivNoExecPerms); err != nil {
+		return "", fmt.Errorf("Error writing public key to location: %s", pubFilePath)
```
Maybe wrap the original error so it's easier to debug? Also errors are supposed to always start with lowercase so that they can be wrapped by other code.

```go
return "', errors.Wrapf(err, "failed to write public key to %s", pubFilePath)
```

dnephin commented on cli/command/trust/key_load.go:
```diff
@@ -0,0 +1,94 @@
+	passRet := trust.GetPassphraseRetriever(streams.In(), streams.Out())
+	keyBytes, err := getPrivKeyBytesFromPath(keyPath)
+	if err != nil {
+		return fmt.Errorf("error reading key from %s: %s", keyPath, err)
```
```go
return errors.Wrapf(err, "error reading key from %s", keyPath)
```

dnephin commented on cli/command/trust/key_load.go:
```diff
@@ -0,0 +1,94 @@
+		return fmt.Errorf("error reading key from %s: %s", keyPath, err)
+	}
+	if err := loadPrivKeyBytesToStore(keyBytes, privKeyImporters, keyPath, options.keyName, passRet); err != nil {
+		return fmt.Errorf("error importing key from %s: %s", keyPath, err)
```
`errors.Wrapf()`

dnephin commented on cli/command/trust/key_load.go:
```diff
@@ -0,0 +1,94 @@
+		return nil, err
+	}
+	if fileInfo.Mode() != ownerReadOnlyPerms && fileInfo.Mode() != ownerReadAndWritePerms {
+		return nil, fmt.Errorf("private key permission from %s should be set to 400 or 600", keyPath)
```
What about 700, or 200? Maybe this could say "private key must not be readable by others" or something like that? 

I think the ssh client says something like this: "It is required that your private key files are NOT accessible by others."

riyazdf replied:
I don't think 200 or 700 make sense here, for 200 you'd have writable but not readable (I've never seen this done?), and with 700 the key would be executable (also don't see a use for this?).

dnephin replied:
I agree those would be silly permissions. My point is that the error message is misleading. The requirement is that it can't be readable by others, not that it must be one of those two permissions. The error message should match the requirement. The ssh client is a good example of this.

If I'm not mistaken this error message will also be used on windows. "set to 400 or 600" is meaningless to windows users.

dnephin replied:
Ah, I just noticed the code *does* actually check for specifically those permissions. I think this needs to be changed to just check that the specific bits are not set.

riyazdf replied:
sounds good, updated!

dnephin commented on cli/command/trust/key_load.go:
```diff
@@ -0,0 +1,94 @@
+	if err != nil {
+		return nil, err
+	}
+	return keyBytes, nil
```
hmm, not sure why gosimple isn't failing here. This should be:

```go
return ioutil.ReadAll(from)
```

dnephin commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,138 @@
+	var errImages []string
+	for _, imageName := range options.images {
+		if err := addSignerToImage(cli, signerName, imageName, options.keys.GetAll()); err != nil {
+			fmt.Fprintln(cli.Out(), err.Error())
```
Should go to `cli.Err()` instead of stdout

dnephin commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,138 @@
+			if os.IsNotExist(err) {
+				return nil, fmt.Errorf("file for public key does not exist: %s", pubKeyPath)
+			}
+			return nil, fmt.Errorf("unable to read public key from file: %s", pubKeyPath)
```
Maybe `errors.Wrapf()` to make it easier to debug with the original error.

dnephin commented on cli/command/trust/signer_remove.go:
```diff
@@ -0,0 +1,137 @@
+	var errImages []string
+	for _, image := range options.images {
+		if err := removeSingleSigner(cli, image, options.signer, options.forceYes); err != nil {
+			fmt.Fprintln(cli.Out(), err.Error())
```
`s/cli.Out()/cli.Err()/`

dnephin commented on cli/command/trust/signer_remove.go:
```diff
@@ -0,0 +1,137 @@
+	}
+	delegationRoles, err := notaryRepo.GetDelegationRoles()
+	if err != nil {
+		return fmt.Errorf("Error retrieving signers for %s", imageName)
```
Maybe `errors.Wrapf()`?

dnephin commented on cli/command/trust/signer_remove.go:
```diff
@@ -0,0 +1,137 @@
+			return nil
+		}
+	} else if err != nil {
+		fmt.Fprintln(cli.Out(), err.Error())
```
Shouldn't this return the error here to exit early?

If not I guess just `s/cli.Out/cli.Err/`

riyazdf commented:
@dnephin: thanks for the quick turnaround!  Addressed all feedback and commented on file perms

thaJeztah commented on cli/command/trust/key_generate.go:
```diff
@@ -0,0 +1,131 @@
+func newKeyGenerateCommand(dockerCli command.Streams) *cobra.Command {
+	options := keyGenerateOptions{}
+	cmd := &cobra.Command{
+		Use:   "generate NAME [NAME...]",
```
Usage is incorrect here, because only one `NAME` can be provided

thaJeztah commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,138 @@
+		pubKeyBytes, err := ioutil.ReadFile(pubKeyPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return nil, fmt.Errorf("file for public key does not exist: %s", pubKeyPath)
```
Looks like we don't have to special case `not exist` errors. Standard errors look to provide the right information. When removing the special case, it looks like:

```
unable to read public key from file: /no-such-key: open /no-such-key: no such file or directory

unable to read public key from file: /foo/foobar.pub: open /foo/foobar.pub: permission denied
```

So no `Wrapf()` needed as well, just:

```go
return nil, errors.Wrap(err, "unable to read public key from file")
```

thaJeztah commented on cli/command/trust/key_generate.go:
```diff
@@ -0,0 +1,131 @@
+
+func setupPassphraseAndGenerateKeys(streams command.Streams, opts keyGenerateOptions) error {
+	targetDir := opts.directory
+	// if the target dir is empty, default to CWD
```
This comment is a bit redundant as it's just the code "in words", also leaning towards having the validation in this step, instead of further down in `validateAndGenerateKey()`, something like:

```patch
diff --git a/cli/command/trust/key_generate.go b/cli/command/trust/key_generate.go
index 913083be..2cb00401 100644
--- a/cli/command/trust/key_generate.go
+++ b/cli/command/trust/key_generate.go
@@ -59,25 +59,23 @@ func validateKeyArgs(keyName string, cwdPath string) error {
 }

 func setupPassphraseAndGenerateKeys(streams command.Streams, opts keyGenerateOptions) error {
-       targetDir := opts.directory
-       // if the target dir is empty, default to CWD
-       if targetDir == "" {
+       if opts.directory == "" {
                cwd, err := os.Getwd()
                if err != nil {
                        return err
                }
-               targetDir = cwd
+               opts.directory = cwd
        }
-       return validateAndGenerateKey(streams, opts.name, targetDir)
-}
-
-func validateAndGenerateKey(streams command.Streams, keyName string, workingDir string) error {
-       freshPassRetGetter := func() notary.PassRetriever { return trust.GetPassphraseRetriever(streams.In(), streams.Out()) }
-       if err := validateKeyArgs(keyName, workingDir); err != nil {
+       if err := validateKeyArgs(opts.name, opts.directory); err != nil {
                return err
        }
+       return generateKey(streams, opts.name, opts.directory)
+}
+
+func generateKey(streams command.Streams, keyName string, workingDir string) error {
        fmt.Fprintf(streams.Out(), "\nGenerating key for %s...\n", keyName)
        // Automatically load the private key to local storage for use
+       freshPassRetGetter := func() notary.PassRetriever { return trust.GetPassphraseRetriever(streams.In(), streams.Out()) }
        privKeyFileStore, err := trustmanager.NewKeyFileStore(trust.GetTrustDirectory(), freshPassRetGetter())
        if err != nil {
                return err
```

thaJeztah commented:
I also see there's still quite a few places that print "informational" information on `stdout`; that may be problematic if people want to script these things (perhaps not a concern, given that passwords have to be entered interactively)

riyazdf commented on cli/command/trust/helpers.go:
```diff
@@ -31,3 +32,13 @@ func clearChangeList(notaryRepo client.Repository) error {
+	if err != nil {
+		return err
+	}
+	// Initialize the notary repository with a remotely managed snapshot
```
added to the signature above

riyazdf commented on cli/command/trust/key_generate.go:
```diff
@@ -0,0 +1,131 @@
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(streams.Out(), "Successfully generated and loaded private key. Corresponding public key available: %s\n", writtenPubFile)
```
I don't want to expose the private key storage to users so I'd prefer not to (similar to comment below on the `TrustDirectory`)

riyazdf commented:
@thaJeztah: thank you for the thorough review!  I think I've addressed all comments.

As for the bug you found: I am not able to repro so maybe it is an odd tty issue in the container?

riyazdf commented on cli/command/trust/key_generate.go:
```diff
@@ -0,0 +1,131 @@
+		},
+	}
+	flags := cmd.Flags()
+	flags.StringVarP(&options.directory, "dir", "d", "", "Directory to generate key in, defaults to current directory")
```
I removed `-d` and left `--dir`, because I thought it'd good to be prescriptive with the filenames themselves (having `.pub` extension and the specified).

I'm holding off on `stdin` for now because the env var `DOCKER_CONTENT_TRUST_REPOSITORY_PASSPHRASE` can be used

tych0 replied:
I see this env variable is from previous commits, so maybe it doesn't need fixing here, but it's worth noting that this is also very unsafe. Just as `/proc/self/cmdline` has the contents of the command line arguments, `/proc/self/environ` has the contents of the task's initial environment, and is world readable, exposing you to similar issues as `/proc/self/cmdline`.

riyazdf commented on cli/command/trust/key_generate.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// key names can use alphanumeric + _ + - characters
+var validKeyName = regexp.MustCompile(`^[a-zA-Z0-9\_]+[a-zA-Z0-9\_\-]*$`).MatchString
```
nice catch on the Mac/Windows and Linux!  I've changed to only lowercase

riyazdf commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,138 @@
+func newSignerAddCommand(dockerCli command.Cli) *cobra.Command {
+	var options signerAddOptions
+	cmd := &cobra.Command{
+		Use:   "add [OPTIONS] NAME IMAGE [IMAGE...] ",
```
We had the flag for `--name` originally but decided from user feedback (and @mbentley and @pdevine) that the positional argument made more sense

riyazdf commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,138 @@
+	}
+	flags := cmd.Flags()
+	options.keys = opts.NewListOpts(nil)
+	flags.VarP(&options.keys, "key", "k", "Path to the signer's public key(s)")
```
removed shorthand `-k`

riyazdf commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,138 @@
+	pubKeys := []data.PublicKey{}
+	for _, pubKeyPath := range pubKeyPaths {
+		// Read public key bytes from PEM file
+		pubKeyBytes, err := ioutil.ReadFile(pubKeyPath)
```
nice catch, limited!

dnephin commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,141 @@
+	for _, imageName := range options.images {
+		fmt.Fprintf(cli.Out(), "Adding signer \"%s\" to %s...\n", signerName, imageName)
+		if err := addSignerToImage(cli, signerName, imageName, signerPubKeys); err != nil {
+			fmt.Fprintln(cli.Err(), err.Error()+"\n")
```
`Fprintln()` already adds a newline, but there's also a `\n` at the end of the message. Are you sure we want an extra blank line between these?

riyazdf replied:
I added an extra blank line to separate the signer output if there are multiple ex:
```
$ docker trust signer add alice example/trust-demo example/trust-demo2 --key alice.crt
Adding signer "alice" to example/trust-demo...
Enter passphrase for repository key with ID 95b9e55: 
Successfully added signer: alice to example/trust-demo

Adding signer "alice" to example/trust-demo2...
Enter passphrase for repository key with ID ece554f: 
Successfully added signer: alice to example/trust-demo2
```

Let me know if you'd prefer for the blank line to be removed, but I think it helps improve readability

dnephin replied:
Ok

dnephin commented on cli/command/trust/signer_remove.go:
```diff
@@ -0,0 +1,138 @@
+	for _, image := range options.images {
+		fmt.Fprintf(cli.Out(), "Removing signer \"%s\" from %s...\n", options.signer, image)
+		if err := removeSingleSigner(cli, image, options.signer, options.forceYes); err != nil {
+			fmt.Fprintln(cli.Err(), err.Error()+"\n")
```
same here

dnephin commented on cli/command/trust/signer_remove.go:
```diff
@@ -0,0 +1,138 @@
+			fmt.Fprintln(cli.Err(), err.Error()+"\n")
+			errImages = append(errImages, image)
+		} else {
+			fmt.Fprintf(cli.Out(), "Successfully removed %s from %s\n\n", options.signer, image)
```
same here, seems to be an extra blank line

thaJeztah commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,141 @@
+	var errImages []string
+	for _, imageName := range options.images {
+		fmt.Fprintf(cli.Out(), "Adding signer \"%s\" to %s...\n", signerName, imageName)
+		if err := addSignerToImage(cli, signerName, imageName, signerPubKeys); err != nil {
```
Not a blocker, but noticed that we're not consistent in requiring a `:tag` or not;

```bash
$ docker trust sign thajeztah/signingtest
No tag specified for thajeztah/signingtest
```

But

```bash
$ docker trust signer add bob thajeztah/signingtest --key=/foo/baz/bob.pub

Adding signer "bob" to thajeztah/signingtest...
Enter passphrase for repository key with ID 05cab95:
Successfully added signer: bob to thajeztah/signingtest
```

riyazdf replied:
for signing a tag is required to specify an image.  For `signer add` and `signer remove` we silently parse out any tags because they operate over the full image repo - we did this for convenience, but we could error out explicitly if you prefer?

thaJeztah replied:
Ah, right, so

```
docker trust signer add bob thajeztah/signingtest:one thajeztah/signingtest:two
```

is basically doing

```
docker trust signer add bob thajeztah/signingtest:*
```

So `docker trust signer add` adds a signer to a _repository_, not an _image_ ? I know "repository" is a bit overloaded, but perhaps we need to use that here (again: could use some input on that one)

riyazdf replied:
yes it is, great catch! Notary operates a single trusted collection per repository. I'll update usage and argnames accordingly

thaJeztah commented on docs/reference/commandline/trust_signer_remove.md:
````diff
@@ -0,0 +1,171 @@
+# trust signer remove
+
+```markdown
+Usage:  docker trust signer remove [OPTIONS] NAME IMAGE [IMAGE...]
````
```
Usage:	docker trust signer remove [OPTIONS] NAME IMAGE [IMAGE...]

Remove a signer

Options:
      --help   Print usage
  -y, --yes    Do not prompt for confirmation before removing the most recent signer
```

thaJeztah commented on docs/reference/commandline/trust_signer_remove.md:
````diff
@@ -0,0 +1,171 @@
+To remove an existing signer, `alice`, from this repository: 
+
+```bash
+$ docker trust inspect example/trust-demo
````
s/inspect/view/

thaJeztah commented on docs/reference/commandline/trust_signer_remove.md:
````diff
@@ -0,0 +1,171 @@
+To remove an existing signer, `alice`, from multiple repositories: 
+
+```bash
+$ docker trust inspect example/trust-demo
````
s/inspect/view/

thaJeztah commented on cli/command/trust/signer_remove.go:
```diff
@@ -0,0 +1,138 @@
+
+func removeSigner(cli command.Cli, options signerRemoveOptions) error {
+	var errImages []string
+	for _, image := range options.images {
```
Looks like we don't require a `:tag` for these images as well

thaJeztah commented on docs/reference/commandline/trust_signer_add.md:
````diff
@@ -0,0 +1,210 @@
+To add a new signer, `alice`, to this repository: 
+
+```bash
+$ docker trust inspect example/trust-demo
````
s/inspect/view/

(some other places as well)

thaJeztah commented on cli/command/trust/signer_remove.go:
```diff
@@ -0,0 +1,138 @@
+		},
+	}
+	flags := cmd.Flags()
+	flags.BoolVarP(&options.forceYes, "yes", "y", false, "Do not prompt for confirmation before removing the most recent signer")
```
Noticed that for `docker xx prune`, we're using `--force` for this;

```
Usage:	docker image prune [OPTIONS]

Remove unused images

Options:
  -a, --all             Remove all unused images, not just dangling ones
      --filter filter   Provide filter values (e.g. 'until=<timestamp>')
  -f, --force           Do not prompt for confirmation
      --help            Print usage
```

riyazdf replied:
I'll update to force :+1:

riyazdf commented on cli/command/trust/key_load.go:
```diff
@@ -0,0 +1,90 @@
+		return fmt.Errorf("provided file %s is not a supported private key - to add a signer's public key use docker trust signer add", keyPath)
+	}
+	// Make a reader, rewind the file pointer
+	return trustmanager.ImportKeys(bytes.NewReader(privKeyBytes), privKeyImporters, keyName, "", passRet)
```
I think this is fine - the name is simply extra metadata attached to the unique ID so it's ok to overwrite it if you have access to the same private key

thaJeztah replied:
For a follow-up, we could add a message printing "updating existing key"

riyazdf commented on cli/command/trust/signer_add.go:
```diff
@@ -0,0 +1,141 @@
+	}
+	flags := cmd.Flags()
+	options.keys = opts.NewListOpts(nil)
+	flags.Var(&options.keys, "key", "Path to the signer's public key file(s)")
```
I think we have a similar `(s)` for options like `volumes-from` so we had it here for consistency - is this okay or are we trying to remove this?

thaJeztah replied:
`volumes-from` is slightly different, because it copies all defined volumes from another container, _and_ can be specified multiple times (so a single `--volumes-from` can lead to one or more volumes being added)

thaJeztah commented on cli/command/trust/key_load.go:
```diff
@@ -82,9 +87,29 @@ func getPrivKeyBytesFromPath(keyPath string) ([]byte, error) {
 	return trustmanager.ImportKeys(bytes.NewReader(privKeyBytes), privKeyImporters, keyName, "", passRet)
 }
+
+func decodePrivKeyIfNecessary(privPemBytes []byte, passRet notary.PassRetriever) ([]byte, error) {
```
Would be good to have a unit-test for this one, but I'm okay adding that in a follow-up

riyazdf commented:
thanks @thaJeztah: I've cut down to ~half the commits!

thaJeztah commented:
Thanks! Discussed with @riyazdf and further squashing proved to be difficult, so I'm good with the current number of commits 👍
